### PR TITLE
Auto mode handling UX improvements

### PR DESCRIPTION
**Pull request by LorenzMeier**
This adds a few useful changes:

  * If you move in LAND, AUTO or HOLD the sticks the system will give control back to the pilot
  * If you do not connect any RC the system will default to HOLD and will allow you tablet control
  * If you gain position lock for the first time the system will re-evaluate the mode switch (so if you dropped down to alt hold it will now go into position)
  * If the system breaches the Geofence it will now always drop back to POSCTRL if the sticks are moved

### CODE REVIEW

**Comment by LorenzMeier** [comment]
@tumbili @dagar @MaEtUgR comments welcome. I think this is all common sense enough.

File: src/modules/events/mag_calibration/task.cpp [hunk 1]
```diff
@@ -0,0 +1,324 @@
+/****************************************************************************
```
Comment (dagar) [inline src/modules/events/mag_calibration/task.cpp]: I think this slipped in unintentionally?

File: src/modules/commander/commander.cpp [hunk 2]
```diff
@@ -2498,15 +2506,33 @@ int commander_thread_main(int argc, char *argv[])
...
 
-				main_state_transition(&status, main_state_before_rtl, main_state_prev, &status_flags, &internal_state);
+				// revert to position control in any case
+				main_state_transition(&status, commander_state_s::MAIN_STATE_POSCTL, main_state_prev, &status_flags, &internal_state);
+				mavlink_log_critical(&mavlink_log_pub, "Aborted RTL, returned control to pilot");
+			}
+		}
+
+		// abort landing or auto or loiter if sticks are moved significantly
+		if (internal_state.main_state == commander_state_s::MAIN_STATE_AUTO_LAND ||
```
Comment (dagar) [inline src/modules/commander/commander.cpp]: Any reason this wasn't combined with the same behaviour for RTL above?

File: src/modules/commander/commander.cpp [hunk 3]
```diff
@@ -2498,15 +2506,33 @@ int commander_thread_main(int argc, char *argv[])
...
+			internal_state.main_state == commander_state_s::MAIN_STATE_AUTO_MISSION ||
+			internal_state.main_state == commander_state_s::MAIN_STATE_AUTO_LOITER) {
+			// transition to previous state if sticks are touched
+			if ((_last_sp_man.timestamp != sp_man.timestamp) &&
+				((fabsf(sp_man.x) - fabsf(_last_sp_man.x) > min_stick_change) ||
+				 (fabsf(sp_man.y) - fabsf(_last_sp_man.y) > min_stick_change) ||
```
Comment (dagar) [inline src/modules/commander/commander.cpp]: Should min_stick_change factor in time between man sp changes?

Comment (LorenzMeier) [inline src/modules/commander/commander.cpp]: I think it should. We should implement a change rate.

File: src/modules/commander/commander.cpp (hunk 2 above)
Comment (LorenzMeier) [inline src/modules/commander/commander.cpp]: Not really. We should do a function which does this as a function of change rate.

**Comment by dagar** [comment]
Thanks for taking care of this, I think it makes a lot of sense.

File: src/modules/commander/commander.cpp [hunk 4]
```diff
@@ -194,6 +194,7 @@ static float max_ekf_dang_bias = 3.5e-4f;
 /* pre-flight IMU consistency checks */
 static float max_imu_acc_diff = 0.7f;
 static float max_imu_gyr_diff = 0.09f;
+static float min_stick_change = 0.25f;
```
Comment (LorenzMeier) [inline src/modules/commander/commander.cpp]: This is not a good default.

File: src/modules/commander/commander.cpp (hunk 3 above)
Comment (LorenzMeier) [inline src/modules/commander/commander.cpp]: Actually - I think last_sp_man is implemented as the last setpoint in the last mode (implicitly the last manual mode). Its where you left the sticks when you handed over control. Locking that position and comparing it makes more sense than a change rate.

**Comment by dagar** [comment]
We might need to play with it to see if it actually matters, but I'm wondering about the case where you're flying manually, violate the fence, and RTL is triggered. Letting the sticks settle back to neutral shouldn't be enough to change the state to POSCTL.

**Comment by LorenzMeier** [comment]
That has been a corner case before as well. I think any pilot would try to take over again. We need to test it. To me RTL on fence breach is sort of incompatible with manual input anyway. Its very scary and counter-intuitive. So maybe breaking out of it immediately is not so bad (if you are inside the radius again).

File: src/modules/commander/commander.cpp (hunk 4 above)
Comment (LorenzMeier) [inline src/modules/commander/commander.cpp]: I recant, that's actually an OK default.

**Comment by AndreasAntener** [comment]
IMHO it comes down to the question if the operator is aware that RTL on fence breach is activated and that a fence is active in the first place. If we assume that is the case, then RTL on fence breach is fine even in manual mode. My suggestion to improve the awareness was to let RTL turn the vehicle around first before you let the user switch out (which would also make sure it's inside the fence again).

**Comment by Stifael** [comment]
I just quickly tested it in sitl: 
moving stick during hold brings me back to position
moving stick during mission brings me back to position
moving stick during rtl  nothing happens.

**Comment by Stifael** [comment]
tested on real vehicle.
It seems that the stick input is not correctly interpreted:
if  in hold:
- there is no switch when moving sticks slowly
- there is also sometimes no switch when moving sticks quickly, but not to the maximum range
if in rtl:
- nothing happens

What exactly is the main advantage of this feature? for instance, if I am in hold and move the sticks, then I would enter position control but the remote controller still has the stick at hold. If I now want to switch back again to hold, then I first have to switch back the stick to position and then back to hold.

File: src/modules/commander/commander.cpp [hunk 5]
```diff
@@ -2498,15 +2506,34 @@ int commander_thread_main(int argc, char *argv[])
 			main_state_before_rtl == commander_state_s::MAIN_STATE_RATTITUDE ||
 			main_state_before_rtl == commander_state_s::MAIN_STATE_STAB)) {
 
-			// transition to previous state if sticks are increased
-			const float min_stick_change = 0.2f;
+			// transition to previous state if sticks are touched
 			if ((_last_sp_man.timestamp != sp_man.timestamp) &&
 				((fabsf(sp_man.x) - fabsf(_last_sp_man.x) > min_stick_change) ||
```
Comment (MaEtUgR) [inline src/modules/commander/commander.cpp]: If I read it correctly this is handled like a switch transition, meaning the value is compared to what it was when the fc last received one. So the actual speed with which you have to move your stick depends on the RC update rate and if you happen to have a high update frequency you might never reach such a step.

**Comment by MaEtUgR** [comment]
I guess RTL is not bothered because you might accidentally overlook an automatic RTL on low battery. But the same would also apply to the land mode. An automatic emergency land because of critial battery could go unnoticed. I know it makes things more complicated but maybe we need a distinction of forced reactions or a way the user can not accidentally switch out of these modes.

**Comment by LorenzMeier** [comment]
We should do this UX changes one release later.

**Comment by dagar** [comment]
Is this still wanted?

**Comment by mhkabir** [comment]
I like this as a feature. Seems very sensible.

**Comment by mcsauder** [comment]
Hi @LorenzMeier and @dagar , I've checked through this PR and the only changes proposed here that are not already incorporated into current PX4:master involve `position_lock_gained`.  Everything else in this PR is already in place in the master branch.

If you would like, I could submit a PR with the `position_lock_gained` logic changes from this PR and this PR could be closed, or this PR could just be closed out.

Let me know if you'd like those remaining changes submitted or how else I can help.

**Comment by Antiheavy** [comment]
> If the system breaches the Geofence it will now always drop back to POSCTRL if the sticks are moved

@mcsauder I recall a discussion within the past few months about how this specific behavior would be inappropriate for Fixed Wing as POSCTRL for fixed wings isn't "just hover" like it is for multicopter.  In your review of this old PR, did you see that feature as MC only?

**Comment by mcsauder** [comment]
Hi @Antiheavy , I think your discussion is unrelated to this particular change.  For convenience here is a compare of what I *believe is the appropriate application of the intent in this PR: https://github.com/PX4/Firmware/compare/master...mcsauder:mode_switch_automodes

I did a quick flight test to see if anything was obviously broken by these changes and things appear to be functioning properly.  (They will require some more scrutiny.)

**Comment by julianoes** [comment]
Remaining diff is here:  #12769 thanks @mcsauder
